Tutorial part ii: platformer (#1021)
agbrs/agb: book/src/platformer/03_importing_tiled_levels.md

# Importing Tiled levels

What we want to write is something that parses the Tiled level and creates a representation in Rust.
This will be run at build time.
We do this at build time to avoid doing any parsing on the GBA itself.

There is a Rust crate for parsing Tiled levels called [`tiled`](https://crates.io/crates/tiled).
To easily write out the Rust code for the levels, we will use the [`quote`](https://crates.io/crates/quote) crate.
In order to have access to the `TokenStream` type, we will also need [`proc-macro2`](https://crates.io/crates/proc-macro2).
Include this in your `Cargo.toml`

```toml
[build-dependencies]
quote = "1"
proc-macro2 = "1"
tiled = "0.14.0"
```

# `build.rs`

> The `build.rs` file placed in the manifest directory will be run before the compilation of your game.

Reviewer: Maybe something along the lines of 'same directory as the Cargo.toml' rather than 'manifest directory'?

impl ResourceReader for BuildResourceReader {
    type Resource = <FilesystemResourceReader as ResourceReader>::Resource;

    type Error = <FilesystemResourceReader as ResourceReader>::Error;

    fn read_from(
        &mut self,
        path: &std::path::Path,
    ) -> std::result::Result<Self::Resource, Self::Error> {
        println!("cargo::rerun-if-changed={}", path.to_string_lossy());
        FilesystemResourceReader.read_from(path)
    }
}
```

This adds a reader that passes through to the existing `FilesystemResourceReader` but intercepts each one to tell `cargo` to depend on the file being read.
The whole reason for doing this is that loading a tiled map could mean we also need to load the various files it references, like the tilesets.
If we were to change the tileset, maybe adding tiles or changing the tags, we would like that to be reflected in the next build of our game.
Make sure to properly tell `cargo` about your dependencies as it will annoy you otherwise!

# Loading the level into an internal representation

It is best practice and easier to maintain to import the level into an internal representation and then convert that into your game representation.

> The internal representation can be inefficient as it's going to use your powerful build machine rather than the underpowered GBA.

Reviewer: What do you mean by this?

Author: There are 4 possible tiles, how I have my intermediate representation inline the collision and win as separate bools is wildly inefficient.

fn main() -> Result<(), Box<dyn Error>> {
    // the file we output to which is in the `OUT_DIR` directory. This is a directory
    // provided by cargo designed to be used by the build script to include output into
    let mut file =
        std::fs::File::create(format!("{}/levels.rs", std::env::var("OUT_DIR").unwrap()))?;

    // make and write each level to the output
    for (number, level) in LEVELS.iter().enumerate() {
        let ident = quote::format_ident!("LEVEL_{}", number);
        let level = import_level(&format!("tiled/{level}"))?;
        let content = quote! {
            static #ident: Level = #level;
        };
        writeln!(file, "{content}")?;
    }

    // define an array of all the levels to be used by the game, therefore make the array `pub`.
    let levels = (0..LEVELS.len()).map(|x| quote::format_ident!("LEVEL_{}", x));
    writeln!(
        file,
        "{}",
        quote! {
            pub static LEVELS: &[&Level] = &[#(&#levels),*];
        }
    )?;

    Ok(())
}
```

# Summary

Here we've made a level import system for our game.
We've shown how you can use the basic Tiled features along with a more advanced one (custom properties).

> How you use tiled is highly individualised and as your projects grow so too will your use of tiled and the various features you use and have to write support for in your import system.
> To make levels quickly and efficiently, you will want to individualise it and make it work for you.

Reviewer: ```suggestion
How you use tiled is very specific and as your projects grow so too will your use of tiled and its various features. 
To make levels quickly and efficiently, you will want to specialise it and make it work for you.
```

Author: I'd like to convey that your use of tiled and your import system are linked. Our use of tiled in Khiera and DPL are completely different and have very different tooling.
